src/goto-symex/symex_goto.cpp

```diff
@@ -129,7 +128,11 @@ static optionalt<renamedt<exprt, L2>> try_evaluate_pointer_comparison(
         value_set_dereferencet::build_reference_to(
           value_set_element, symbol_expr, ns);
 
-      if(skip_typecast(value.pointer) == other_without_typecast)
+      // use the simplifier to test equality as we need to skip over typecasts
+      // and cannot rely on canonical representations, which would permit a
+      // simple syntactic equality test
+      equal_exprt test_equal{value.pointer, other_operand};
+      if(simplify_expr(test_equal, ns).is_true())
```

Comment:
This code assumes that if this resolves to true, then it's a match, which is fine.

But it also assumes that if it does not resolves to true, then it's not a match, which is not entirely correct. If it does not resolve to true, it does not mean it's false. It could mean it's to complex to evaluate.

Does that make sense? Perhaps one should add:
```
    else if (!simplified_condition.is_false()) { return {}; /* cannot say anything */ }
```

Reply:
You're of course right. I have amended the code, and also had to make some further fixes (mainly in the simplifier).